### PR TITLE
testdrive: clean up kafka state at beginning of test script

### LINKED ISSUE #15090: cloudtest: remove testdrive hack for controlling environment resets
#15088 included a hack to avoid tearing down Materialize state until the end of execution.

Currently the reset mechanism controls both the Materialize environment on initialization (i.e. clearing the catalog), as well as cleaning up the environment to try and leak as few resources as possible.

We could control this with other flags (reset_mz, reset_env), and let the current reset option imply both.

### PR DESCRIPTION
**Pull request by benesch**
@sploiselle I thiiiiink this fixes #15090. We'll see what CI says.

----

The testdrive philosophy is to leave around state at the end of the
execution. Instead cleanup of the *last* test script is performed at the
beginning of each test run.  This makes it easy to debug a testdrive
script, because the state is left around after testdrive exits.  It also
makes it possible to use a testdrive script in demos or development as
to set up a Materialize instance for further tinkering.

The one wrinkle with this approach is that Kafka is not always
available. So condition the cleanup of Kafka state on whether the test
script uses a Kafka-related action.

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: src/testdrive/src/lib.rs [hunk 1]
```diff
@@ -80,11 +81,30 @@ async fn run_line_reader(
...
+        )
+    });
+
     let (mut state, state_cleanup) = action::create_state(config).await?;
     let actions = action::build(cmds, &state).await?;
 
     if config.reset {
+        // Delete any existing Materialize and Kafka state *before* the test
+        // script starts. We don't clean up Materialize or Kafka state at the
+        // end of the script because.
```
Comment (sploiselle) [inline src/testdrive/src/lib.rs:98]: "because." 😂 when you don't need to justify yourself to anyone

File: src/testdrive/src/lib.rs [hunk 2]
```diff
@@ -96,11 +116,13 @@ async fn run_line_reader(
     }
 
     if config.reset {
-        let mut errors = Vec::new();
+        // Clean up AWS state at the end of the run. Unlike Materialize and
```
Comment (sploiselle) [inline src/testdrive/src/lib.rs:119]: ty for this explanation; I had missed/misunderstood this context

File: src/testdrive/src/lib.rs [hunk 3]
```diff
@@ -96,11 +116,13 @@ async fn run_line_reader(
...
+        // Kafka state, leaving around AWS resources costs real money. We
+        // intentionally don't stop at the first error because we don't want
+        // to e.g. skip cleaning up SQS resources because we failed to clean up
+        // S# resources.
```
Comment (sploiselle) [inline src/testdrive/src/lib.rs:123]: nit: S3

**Review by sploiselle — APPROVED** [review]
LGTM mod unfinished sentence

File: src/testdrive/src/lib.rs (hunk 2 above)
Comment (benesch) [reply src/testdrive/src/lib.rs:121]: No no, it wasn't written anywhere. Just implicitly encoded across a bunch of testdrive scripts and compositions. 🙈

**Comment by benesch** [comment]
Oh boy, I must have been more tired than I realized last night. Thanks for catching all those typos. All fixed, and marked for auto merge.

**Pull request merged**